llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+			assert(start->getParent() == data.getFunction());
+	}
+	BasicBlockGroupData(FunctionData& data)
+		: BasicBlockGroupData(data, /*parentBBGD*/nullptr, getAllBasicBlocks(*data.getFunction()), &data.getFunction()->getEntryBlock())
+	{
+	}
+	BasicBlockGroupData(BasicBlockGroupData& BBGData)
+		: data(BBGData.data), inner(BBGData.inner), start(nullptr), from(nullptr), parent(&BBGData)
+	{
+	}
+	void addIncomingEdge(const llvm::BasicBlock* comingFrom, const llvm::BasicBlock* target)
+	{
+		assert(inner.count(target));
+		if (start == nullptr)
+		{
+			start = target;
+			from = comingFrom;
+		}
+		if (start != target)
+		{
+			multiHead = true;
+			from = nullptr;
+		}
+		if (comingFrom != from)
+		{
+			from = nullptr;
+		}
+	}
+	void runVisitBasicBlock(const llvm::BasicBlock& BB, std::vector<const llvm::BasicBlock*>& visitNext)
+	{
+		assert(visitNext.empty());
+
+		data.currentEE->setIncomingBB(from);
+		const BasicBlock* ret = data.currentEE->visitBasicBlock<PartialInterpreter::VisitingPolicy::COMPUTE_VALUES>(&BB, from);
+
+		if (ret)
+		{
+			visitNext.push_back(ret);
+		}
+		else
+		{
+			llvm::DenseSet<const llvm::BasicBlock*> setSuccessors;
+			for (auto* bb : successors(&BB))
+			{
+				if (setSuccessors.insert(bb).second)
+					visitNext.push_back(bb);
+			}
+		}
+		//Do the visit of the BB, with comingFrom (possibly nullptr) as predecessor
+		//Loop backs will be directed to another BBgroup
+		//The visit will return the set of reachable BBs -> add them to the subGroups data accordingly
+	}
+	bool shouldCleanUp()
+	{
+		for (auto& p : escaping)
+		{
+			//llvm::errs() <<"escaping:\t" << p.first->getName() << "\t" << p.second <<"\n";
+			if (data.visitCounter[p.first] != p.second)
+				return true;
+		}
+		return false;
+	}
+	bool notifySuccessor(const llvm::BasicBlock* from, const llvm::BasicBlock* succ, const int iter)
+	{
+		if (inner.count(succ) == 0)
+		{
+			assert(parent);
+			//It should handled by the parent SCC
+
+			const bool notified = parent->notifySuccessor(from, succ, iter);
+
+			if (notified && !visitingAll)
+			{
+				escaping.push_back({from, iter});
+			}
+
+			return false;
+		}
+		else if (visitingAll)
+		{
+			return false;
+		}
+		else
+		{
+			int X = -subGroupsID.at(succ);
+			assert( X < (int)subGroups.size());
+			subGroups.at(X).addIncomingEdge(from, succ);
+
+			return true;
+		}
+	}
+	void visitAll()
+	{
+		visitingAll = true;
+		for (const llvm::BasicBlock* bb : inner)
+		{
+			data.currentEE->setIncomingBB(nullptr);
+			data.currentEE->visitBasicBlock<PartialInterpreter::VisitingPolicy::REMOVE_VALUES>(bb, nullptr);
+			for (const llvm::BasicBlock* succ : successors(bb))
+				registerEdge(bb, succ);
+		}
+	}
+	void registerEdge(const llvm::BasicBlock* from, const llvm::BasicBlock* to)
+	{
+		data.registerEdge(from, to);
+		
+		notifySuccessor(from, to, iterationCount);
+	}
+	void recursiveVisit()
+	{
+		if (multiHead)
+		{
+			//Mark everything as reachable
+			visitAll();
+			return;
+		}
+		if (!start)
```

Comment:
How can this happen?

Reply:
Added comment, basically say START if (something)-> A else B, and you can prove it's always A, then B will be visited but start will still nullptr. (Here START, A and B are generic SCC, not just blocks)